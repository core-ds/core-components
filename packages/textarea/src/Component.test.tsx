import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Textarea } from './index';

describe('Textarea', () => {
    describe('Snapshots tests', () => {
        it('should match snapshot', () => {
            const { container } = render(<Textarea value='value' nativeScrollbar={true} />);

            expect(container).toMatchSnapshot();
        });

        it('should match snapshot with default counter', () => {
            const { container } = render(
                <Textarea
                    value='value'
                    maxLength={500}
                    nativeScrollbar={true}
                    showCounter={true}
                />,
            );

            expect(container).toMatchSnapshot();
        });

        it('should match snapshot with custom counter', () => {
            const { container } = render(
                <Textarea
                    nativeScrollbar={true}
                    value='value'
                    maxLength={500}
                    showCounter={true}
                    getCounterText={() => 'Custom counter'}
                />,
            );

            expect(container).toMatchSnapshot();
        });

        it('should match snapshot with custom scrollbar', () => {
            const { container } = render(<Textarea nativeScrollbar={false} value='value' />);

            expect(container).toMatchSnapshot();
        });
    });

    it('should forward ref to textarea', () => {
        const textareaRef = jest.fn();
        const dataTestId = 'test-id';
        render(<Textarea ref={textareaRef} dataTestId={dataTestId} />);

        expect(textareaRef.mock.calls).toEqual([[screen.getByTestId(dataTestId)]]);
    });

    it('should set `data-test-id` attribute to textarea', () => {
        const dataTestId = 'test-id';
        render(<Textarea block={true} dataTestId={dataTestId} />);

        expect(screen.getByTestId(dataTestId).tagName).toBe('TEXTAREA');
    });

    describe('Classes tests', () => {
        it('should set `className` class to form-control wrapper', () => {
            const { container } = render(<Textarea className='test-class' />);

            expect(container.querySelector('.test-class')).toHaveClass('component');
        });

        it('should set `textareaClassName` class to textarea', () => {
            const className = 'test-class';
            render(<Textarea textareaClassName={className} />);

            expect(screen.getByRole('textbox')).toHaveClass(className);
        });

        describe('when component is controlled', () => {
            it('should set `filled` class when value passed', () => {
                render(<Textarea value='some value' nativeScrollbar={true} />);

                expect(screen.getByRole('textbox')).toHaveClass('filled');
            });

            it('should not set `filled` class if the value is empty', () => {
                render(<Textarea value='' />);

                expect(screen.getByRole('textbox')).not.toHaveClass('filled');
            });

            it('should unset `filled` class if the value becomes empty', () => {
                const { rerender } = render(<Textarea value='some value' />);

                rerender(<Textarea value='' />);

                expect(screen.getByRole('textbox')).not.toHaveClass('filled');
            });
        });

        describe('when component is uncontrolled', () => {
            it('should set `filled` class when defaultValue passed', () => {
                render(<Textarea defaultValue='some value' nativeScrollbar={true} />);

                expect(screen.getByRole('textbox')).toHaveClass('filled');
            });

            it('should not set `filled` class if the value is empty', () => {
                render(<Textarea nativeScrollbar={true} />);

                expect(screen.getByRole('textbox')).not.toHaveClass('filled');
            });

            it('should unset `filled` class if value becomes empty', async () => {
                render(<Textarea defaultValue='some value' nativeScrollbar={true} />);

                const textarea = screen.getByRole('textbox') as HTMLInputElement;

                await userEvent.type(textarea, '{backspace}', {
                    initialSelectionStart: 0,
                    initialSelectionEnd: textarea.value.length,
                });

                fireEvent.blur(textarea);

                expect(textarea.value).toBe('');
                expect(textarea).not.toHaveClass('filled');
            });
        });

        it('should set `hasInnerLabel` class', () => {
            render(<Textarea label='label' />);

            expect(screen.getByRole('textbox')).toHaveClass('hasInnerLabel');
        });

        it('should set `resizeVertical` class', () => {
            render(<Textarea resize='vertical' />);

            expect(screen.getByRole('textbox')).toHaveClass('resizeVertical');
        });

        it('should set `disabled` attribute', () => {
            render(<Textarea disabled={true} />);

            expect(screen.getByRole('textbox')).toHaveAttribute('disabled');
        });

        it('should set `maxHeight` style with autosize on', () => {
            render(<Textarea nativeScrollbar={true} autosize={true} maxHeight={100} />);

            expect(screen.getByRole('textbox')).toHaveStyle('max-height: 100px');
        });

        it('should set `maxHeight` style with autosize off', () => {
            render(<Textarea nativeScrollbar={true} autosize={false} maxHeight={100} />);

            expect(screen.getByRole('textbox')).toHaveStyle('max-height: 100px');
        });
    });

    describe('Callbacks tests', () => {
        it('should call `onChange` prop', () => {
            const cb = jest.fn();
            const value = '123';
            render(<Textarea onChange={cb} />);

            const textarea = screen.getByRole('textbox') as HTMLTextAreaElement;

            fireEvent.change(textarea, { target: { value } });

            expect(cb).toBeCalledTimes(1);
            expect(textarea.value).toBe(value);
        });

        it('should call `onFocus` prop', () => {
            const cb = jest.fn();
            render(<Textarea onFocus={cb} />);

            fireEvent.focus(screen.getByRole('textbox'));

            expect(cb).toBeCalledTimes(1);
        });

        it('should call `onBlur` prop', () => {
            const cb = jest.fn();
            render(<Textarea onBlur={cb} />);

            fireEvent.blur(screen.getByRole('textbox'));

            expect(cb).toBeCalledTimes(1);
        });

        it('should not call `onChange` prop if disabled', () => {
            const cb = jest.fn();
            render(<Textarea onChange={cb} disabled={true} />);

            const textarea = screen.getByRole('textbox') as HTMLTextAreaElement;

            userEvent.type(textarea, '123');

            expect(cb).not.toBeCalled();
        });
    });

    it('should unmount without errors', () => {
        const { unmount } = render(<Textarea value='value' onChange={jest.fn()} />);

        expect(unmount).not.toThrowError();
    });
});
