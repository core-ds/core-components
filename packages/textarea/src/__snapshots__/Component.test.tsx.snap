// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea Snapshots tests should match snapshot 1`] = `
<div>
  <div
    class="component component s"
  >
    <div
      class="inner inner filled"
    >
      <div
        class="inputWrapper"
      >
        <div
          class="input"
        >
          <textarea
            autocomplete="on"
            class="textarea textarea s filled"
            rows="1"
          >
            value
          </textarea>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Textarea Snapshots tests should match snapshot with custom counter 1`] = `
<div>
  <div
    class="component component s"
  >
    <div
      class="inner inner filled"
    >
      <div
        class="inputWrapper"
      >
        <div
          class="input"
        >
          <textarea
            autocomplete="on"
            class="textarea textarea s filled"
            rows="1"
          >
            value
          </textarea>
        </div>
      </div>
    </div>
    <span
      class="sub hint"
    >
      Custom counter
    </span>
  </div>
</div>
`;

exports[`Textarea Snapshots tests should match snapshot with custom scrollbar 1`] = `
<div>
  <div
    class="component component s"
  >
    <div
      class="inner inner filled"
    >
      <div
        class="inputWrapper"
      >
        <div
          class="input"
        >
          <div
            class="component component scrollable s filled"
            data-simplebar="init"
            style="padding: 0px; width: 0px;"
          >
            <div
              class="wrapper"
              style="margin: -0px -0px -0px -0px;"
            >
              <div
                class="heightAutoObserverWrapper"
              >
                <div
                  class="heightAutoObserver"
                />
              </div>
              <div
                class="mask"
                style=""
              >
                <div
                  class="offset"
                  style="right: 0px; bottom: 0px;"
                >
                  <div
                    aria-label="scrollable content"
                    class="contentWrapper"
                    role="region"
                    style="height: auto; overflow-x: hidden; overflow-y: hidden;"
                    tabindex="0"
                  >
                    <div
                      class="content scrollableWrapper"
                      style="padding: 0px 0px 0px 0px;"
                    >
                      <textarea
                        autocomplete="on"
                        class="textarea textarea s"
                        rows="1"
                      >
                        value
                      </textarea>
                      <textarea
                        class="textarea textarea s textareaHidden"
                        rows="1"
                      >
                        value
                      </textarea>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="placeholder"
                style="width: 0px; height: 0px;"
              />
            </div>
            <div
              class="track horizontal"
              style="visibility: hidden;"
            >
              <div
                class="scrollbar"
                style="width: 0px; display: none;"
              />
            </div>
            <div
              class="track vertical"
              style="visibility: hidden;"
            >
              <div
                class="scrollbar"
                style="height: 0px; display: none;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Textarea Snapshots tests should match snapshot with default counter 1`] = `
<div>
  <div
    class="component component s"
  >
    <div
      class="inner inner filled"
    >
      <div
        class="inputWrapper"
      >
        <div
          class="input"
        >
          <textarea
            autocomplete="on"
            class="textarea textarea s filled"
            rows="1"
          >
            value
          </textarea>
        </div>
      </div>
    </div>
    <span
      class="sub hint"
    >
      5/500 символов
    </span>
  </div>
</div>
`;
